src/EditorFeatures/Core.Wpf/Workspaces/WpfTextBufferVisibilityTracker.cs

```diff
@@ -0,0 +1,175 @@
+
+using System;
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using System.ComponentModel.Composition;
+using System.Linq;
+using Microsoft.CodeAnalysis.Editor;
+using Microsoft.CodeAnalysis.Editor.Shared.Utilities;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Editor;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Workspaces
+{
+    [Export(typeof(ITextBufferVisibilityTracker))]
+    internal sealed class WpfTextBufferVisibilityTracker : ITextBufferVisibilityTracker
+    {
+        private readonly ITextBufferAssociatedViewService _associatedViewService;
+        private readonly IThreadingContext _threadingContext;
+
+        private readonly Dictionary<ITextBuffer, VisibleTrackerData> _subjectBufferToCallbacks = new();
+
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public WpfTextBufferVisibilityTracker(
+            ITextBufferAssociatedViewService associatedViewService,
+            IThreadingContext threadingContext)
+        {
+            _associatedViewService = associatedViewService;
+            _threadingContext = threadingContext;
+
+            associatedViewService.SubjectBuffersConnected += AssociatedViewService_SubjectBuffersConnected;
+            associatedViewService.SubjectBuffersDisconnected += AssociatedViewService_SubjectBuffersDisconnected;
+        }
+
+        private void AssociatedViewService_SubjectBuffersConnected(object sender, SubjectBuffersConnectedEventArgs e)
+        {
+            Contract.ThrowIfFalse(_threadingContext.HasMainThread);
+            UpdateAllAssociatedViews(e.SubjectBuffers);
+        }
+
+        private void AssociatedViewService_SubjectBuffersDisconnected(object sender, SubjectBuffersConnectedEventArgs e)
+        {
+            Contract.ThrowIfFalse(_threadingContext.HasMainThread);
+            UpdateAllAssociatedViews(e.SubjectBuffers);
+        }
+
+        private void UpdateAllAssociatedViews(ReadOnlyCollection<ITextBuffer> subjectBuffers)
+        {
+            // Whenever views get attached/detached from buffers, make sure we're hooked up to the appropriate events
+            // for them.
+            foreach (var buffer in subjectBuffers)
+            {
+                if (_subjectBufferToCallbacks.TryGetValue(buffer, out var data))
+                    data.UpdateAssociatedViews();
+            }
+        }
+
+        public bool IsVisible(ITextBuffer subjectBuffer)
+        {
+            Contract.ThrowIfFalse(_threadingContext.HasMainThread);
+
+            var views = _associatedViewService.GetAssociatedTextViews(subjectBuffer).ToImmutableArrayOrEmpty();
+
+            // if any of the views were *not* wpf text views, assume the buffer is visible.  We don't know how to
+            // determine the visibility of this buffer.
+            if (views.Any(v => v is not IWpfTextView))
```

Comment:
what would be a valid case of us using this and having a non wpf view?  Is expected that calling this for a C#/vb/ts/f# buffer would not be?

Just wondering if this a bug in the caller and should throw